Fix Excessive Encoding in Test Logs

ivdiazsa opened the pull request:
Fixes #92092. At some point during the refactoring of the test logs generation systems, we added some encoding to the tests' recorded outputs. However, this encoding replaced every non-alphanumeric character with a hex code, including the most common ones like brackets, parentheses, and spaces. This, in turn, rendered the logs visually unappealing and very difficult to read, which complicated all test investigations. This PR fixes that and makes the logs easily readable again.

ivdiazsa committed:
Removed special encoding that was rendering the test logs near
impossible to read properly.

ivdiazsa commented:
/azp run runtime-coreclr outerloop

trylek reviewed (approved):
Looks great to me, thanks Ivan for fixing this!

ivdiazsa committed:
Adjusted the offending test to print the invalid character's hex code
instead, and fixed it alongside its sibling test because they didn't
handle all correct/incorrect cases properly.

ivdiazsa commented:
/azp run runtime-coreclr outerloop

ivdiazsa committed:
Added special handling for illegal XML characters in the test results'
XML logs.

ivdiazsa commented:
/azp run runtime-coreclr outerloop

danmoseley commented on src/tests/Common/XUnitWrapperLibrary/TestSummary.cs:
```diff
@@ -95,12 +102,66 @@ public string ToXmlString()
+
+        private string SanitizeOutput(string output)
+        {
+            Dictionary<char, string>? invalidChars = FindInvalidXMLChars(output);
```
Will this do something reasonable for surrogate pairs? They tend to appear in test output because they are interesting test inputs.

danmoseley commented on src/tests/Common/XUnitWrapperLibrary/TestSummary.cs:
```diff
@@ -95,12 +102,66 @@ public string ToXmlString()
+            StringBuilder sanitizedOutput = new StringBuilder();
+            foreach (char ch in output)
+            {
+                string? invalidCharHexCode = string.Empty;
```
Why not just do a Replace on the string for each of the problematic characters rather than walking the string yourself? You know the string you are replacing with does not contain any invalid characters

danmoseley replied:
Although I'm not sure why you don't just walk the steering checking each char is valid and replacing it on the fly rather than building up a dictionary

ivdiazsa replied:
> Why not just do a Replace on the string for each of the problematic characters rather than walking the string yourself? You know the string you are replacing with does not contain any invalid characters

`String.Replace()` always creates a new string so it's inefficient to be creating new ones for each problematic character, especially considering some tests might have large outputs, and we have lots of tests to process.

ivdiazsa replied:
> Although I'm not sure why you don't just walk the string checking each char is valid and replacing it on the fly rather than building up a dictionary first?

I did it to just return the output as is, in case there were no problematic characters, but yeah it might be simpler as you suggest. Will try it right now.

danmoseley replied:
> String.Replace() always creates a new string so it's inefficient to be creating new ones for each problematic character

Right, for several replacements you want StringBuilder.Replace.

ivdiazsa committed:
Simplified the sanitizing algorithm to one pass, as per Dan's feedback.
